File: daq/slc/extra/javalib/css/org.csstudio.belle2.daq/src/org/belle2/daq/nsm/NSMVar.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package org.belle2.daq.nsm;

import java.io.IOException;

import org.belle2.daq.base.Reader;
import org.belle2.daq.base.Serializable;
import org.belle2.daq.base.Writer;

/**
 * 
 * @author tkonno
 */
public class NSMVar implements Serializable {

	public static final int NONE = 0;
	public static final int INT = 1;
	public static final int FLOAT = 2;
	public static final int TEXT = 3;

	private String m_node = "";
	private String m_name = "";
	private int m_type = NONE;
	private int[] m_var_i = null;
	private float[] m_var_f = null;
	private String m_var_t = null;
	private int m_len = 0;
	private int m_id = 0;
	private int m_revision = 0;

	public NSMVar() {
	}

	public NSMVar(String name, String value) {
		m_type = TEXT;
		m_name = name;
		m_var_t = value;
		m_len = value.length();
	}

	public NSMVar(String name, int value) {
		m_type = INT;
		m_name = name;
		m_var_i = new int[1];
		m_var_i[0] = value;
		m_len = 0;
	}

	public NSMVar(String name, float value) {
		m_type = FLOAT;
		m_name = name;
		m_var_f = new float[1];
		m_var_f[0] = value;
		m_len = 0;
	}

	public NSMVar(String name, int[] value) {
		m_type = INT;
		m_name = name;
		m_var_i = value;
		m_len = value.length;
	}

	public NSMVar(String name, float[] value) {
		m_type = FLOAT;
		m_name = name;
		m_var_f = value;
		m_len = value.length;
	}

	public NSMVar(String value) {
		this("", value);
	}

	public NSMVar(int value) {
		this("", value);
	}

	public NSMVar(float value) {
		this("", value);
	}

	public NSMVar(int[] value) {
		this("", value);
	}

	public NSMVar(float[] value) {
		this("", value);
	}

	public void setName(String name) {
		m_name = name;
	}

	public void setNode(String node) {
		m_node = node;
	}

	public int size() {
		switch (m_type) {
		case NONE:
			return 0;
		case INT:
			return m_var_f.length * 4;
		case FLOAT:
			return m_var_f.length * 4;
		case TEXT:
			return m_var_t.length();
		}
		return 0;
	}

	public String getNode() {
		return m_node;
	}

	public String getName() {
		return m_name;
	}

	public int getType() {
		return m_type;
	}

	public int getLength() {
		switch (m_type) {
		case INT:
			return (m_var_i != null) ? m_var_i.length : 0;
		case FLOAT:
			return (m_var_f != null) ? m_var_f.length : 0;
		case TEXT:
			return (m_var_t != null) ? m_var_t.length() : 0;
		}
		return 0;
	}

	public int getInt() {
		return (m_type == INT && m_var_i != null && m_len == 0) ? m_var_i[0]
				: 0;
	}

	public float getFloat() {
		return (m_type == FLOAT && m_var_f != null && m_len == 0) ? m_var_f[0]
				: 0;
	}

	public String getText() {
		return (m_type == TEXT && m_var_t != null) ? m_var_t : "";
	}

	public int getInt(int index) {
		return (m_type == INT && m_var_i != null && m_len > index) ? m_var_i[index]
				: 0;
	}

	public float getFloat(int index) {
		return (m_type == FLOAT && m_var_f != null && m_len > index) ? m_var_f[index]
				: 0;
	}

	@Override
	public void readObject(Reader reader) throws IOException {
		m_node = reader.readString();
		m_name = reader.readString();
		m_type = reader.readInt();
		m_len = reader.readInt();
		m_id = reader.readInt();
		m_revision = reader.readInt();
		int length = (m_len > 0) ? m_len : 1;
		m_var_i = null;
		m_var_f = null;
		m_var_t = null;
		switch (m_type) {
		case INT:
			m_var_i = new int[length];
			for (int i = 0; i < length; i++) {
				m_var_i[i] = reader.readInt();
			}
			return;
		case FLOAT:
			m_var_f = new float[length];
			for (int i = 0; i < length; i++) {
				m_var_f[i] = reader.readFloat();
			}
			return;
		case TEXT:
			StringBuilder s = new StringBuilder();
			for (int i = 0; i < length; i++) {
				char c = reader.readChar();
				if (c == '\n' || c == '\r' || (c >= ' ' && c <= '~')) {
					s.append(c);
				}
			}
			m_var_t = s.toString();
		}
	}

	@Override
	public void writeObject(Writer writer) throws IOException {
		writer.writeString(m_node);
		writer.writeString(m_name);
		writer.writeInt(m_type);
		writer.writeInt(m_len);
		writer.writeInt(m_id);
		writer.writeInt(m_revision);
		switch (m_type) {
		case INT:
			for (int v : m_var_i) {
				writer.writeInt(v);
			}
			return;
		case FLOAT:
			for (float v : m_var_f) {
				writer.writeFloat(v);
			}
			return;
		case TEXT:
			for (char v : m_var_t.toCharArray()) {
				writer.writeChar(v);
			}
		}
	}

}

File: daq/slc/nsm/src/NSMVar.cc
/**************************************************************************
 * basf2 (Belle II Analysis Software Framework)                           *
 * Author: The Belle II Collaboration                                     *
 *                                                                        *
 * See git log for contributors and copyright holders.                    *
 * This file is licensed under LGPL-3.0, see LICENSE.md.                  *
 **************************************************************************/
#include "daq/slc/nsm/NSMVar.h"

#include <string.h>
#include <stdlib.h>

#include <daq/slc/base/Reader.h>
#include <daq/slc/base/Writer.h>

using namespace Belle2;

NSMVar::NSMVar(const std::string& name, const std::vector<int>& value)
  : m_value(NULL)
{
  int* v = (int*)malloc(value.size() * sizeof(int));
  for (size_t i = 0; i < value.size(); i++)
    v[i] = value[i];
  m_name = name;
  m_type = INT;
  m_len = value.size();
  m_value = v;
  m_id = 0;
  m_date = Date().get();
}

NSMVar::NSMVar(const std::string& name, const std::vector<float>& value)
  : m_value(NULL)
{
  float* v = (float*)malloc(value.size() * sizeof(float));
  for (size_t i = 0; i < value.size(); i++)
    v[i] = value[i];
  m_name = name;
  m_type = INT;
  m_len = value.size();
  m_value = v;
  m_id = 0;
  m_date = Date().get();
}

const NSMVar& NSMVar::operator=(const std::vector<int>& val)
{
  if (m_value) free(m_value);
  int* v = (int*)malloc(val.size() * sizeof(int));
  for (size_t i = 0; i < val.size(); i++)
    v[i] = val[i];
  m_value = v;
  m_type = INT;
  m_len = val.size();
  m_date = Date().get();
  return *this;
}

const NSMVar& NSMVar::operator=(const std::vector<float>& val)
{
  if (m_value) free(m_value);
  float* v = (float*)malloc(val.size() * sizeof(int));
  for (size_t i = 0; i < val.size(); i++)
    v[i] = val[i];
  m_value = v;
  m_type = FLOAT;
  m_len = val.size();
  m_date = Date().get();
  return *this;
}

int NSMVar::size() const
{
  int s = 1;
  switch (m_type) {
    case NONE: s = 0; return 0;
    case INT: s = 4; break;
    case FLOAT: s = 4; break;
    case TEXT: return m_len + 1;
    default:  break;
  }
  return (m_len > 0) ? s * m_len : s;
}

const char* NSMVar::getTypeLabel() const
{
  switch (m_type) {
    case INT:
      if (m_len > 0) return "inta";
      else return "int";
    case FLOAT:
      if (m_len > 0) return "floata";
      else return "float";
    case TEXT: return "text";
    default:  break;
  }
  return "";
}

void NSMVar::copy(const std::string& name,
                  Type type, int len, const void* value,
                  int id, int date)
{
  if (m_value) free(m_value);
  m_value = NULL;
  m_name = name;
  m_type = type;
  m_len = len;
  m_id = id;
  m_date = (date > 0) ? date : Date().get();
  int s = size();
  if (s > 0) {
    m_value = malloc(s);
    memcpy(m_value, value, s);
  } else {
    m_type = NONE;
    m_len = 0;
    m_value = NULL;
  }
}

NSMVar::~NSMVar()
{
  if (m_value) free(m_value);
}

void NSMVar::readObject(Reader& reader)
{
  m_node = reader.readString();
  m_name = reader.readString();
  m_type = (Type)reader.readInt();
  m_len = reader.readInt();
  m_id = reader.readInt();
  m_date = reader.readInt();
  int len = (m_len > 0) ? m_len : 1;
  int s = size();
  if (s > 0) {
    if (m_value) free(m_value);
    m_value = malloc(s);
    memset(m_value, 0, s);
    switch (m_type) {
      case INT: {
        int* v = (int*)m_value;
        for (int i = 0; i < len; i++) {
          v[i] = reader.readInt();
        }
      } break;
      case FLOAT: {
        float* v = (float*)m_value;
        for (int i = 0; i < len; i++) {
          v[i] = reader.readFloat();
        }
      } break;
      case TEXT: {
        char* v = (char*)m_value;
        // Use m_len condition instead of len due to empty string problem
        for (int i = 0; i < m_len; i++) {
          v[i] = reader.readChar();
        }
      } break;
      default: return;
    }
  }
}

void NSMVar::writeObject(Writer& writer) const
{
  writer.writeString(m_node);
  writer.writeString(m_name);
  writer.writeInt((int)m_type);
  writer.writeInt(m_len);
  writer.writeInt(m_id);
  writer.writeInt(m_date);
  int len = (m_len > 0) ? m_len : 1;
  switch (m_type) {
    case INT: {
      int* v = (int*)m_value;
      for (int i = 0; i < len; i++) {
        writer.writeInt(v[i]);
      }
    } break;
    case FLOAT: {
      float* v = (float*)m_value;
      for (int i = 0; i < len; i++) {
        writer.writeFloat(v[i]);
      }
    } break;
    case TEXT: {
      char* v = (char*)m_value;
      writer.write(v, len);
    } break;
    default: return;
  }
}

int NSMVar::getInt() const
{
  return (m_type == INT&& m_len == 0) ? *(int*)m_value : 0;
}

float NSMVar::getFloat() const
{
  return (m_type == FLOAT && m_len == 0) ? *(float*)m_value : 0;
}

const char* NSMVar::getText() const
{
  return (m_type == TEXT && m_len > 0 && m_value != NULL) ? (const char*)m_value : "";
}

int NSMVar::getInt(int i) const
{
  return (m_type == INT&& i < m_len) ? ((int*)m_value)[i] : 0;
}

float NSMVar::getFloat(int i) const
{
  return (m_type == FLOAT && i < m_len) ? ((float*)m_value)[i] : 0;
}

const NSMVar& NSMVar::operator>>(std::vector<int>& val) const
{
  int* pv = (int*)get();
  val = std::vector<int>();
  for (int i = 0; i < getLength(); i++) {
    val.push_back(pv[i]);
  }
  return *this;
}

const NSMVar& NSMVar::operator>>(std::vector<float>& val) const
{
  float* pv = (float*)get();
  val = std::vector<float>();
  for (int i = 0; i < getLength(); i++) {
    val.push_back(pv[i]);
  }
  return *this;
}
